xsuite/src/proxy/csproxy.ts

```diff
@@ -0,0 +1,171 @@
+import { addressToBechAddress } from '../data/address';
+import { BroadTx, Proxy, unrawTxRes } from './proxy';
+import { e, eCodeMetadata, EncodableAccount } from '../data/encoding';
+import { Kvs } from '../data/kvs';
+
+export class CSProxy extends Proxy {
+  stopChainSimulator: () => void;
+  autoGenerateBlocks: boolean;
+  verbose: boolean;
+
+  constructor(baseUrl: string, stopChainSimulator: () => void, autoGenerateBlocks: boolean = true, verbose: boolean = false) {
+    super(baseUrl);
+
+    this.stopChainSimulator = stopChainSimulator;
+    this.autoGenerateBlocks = autoGenerateBlocks;
+    this.verbose = verbose;
+  }
+
+  static async setAccount(baseUrl: string, account: EncodableAccount, autoGenerateBlocks: boolean = true, verbose: boolean = false) {
+    const [previousAccount, previousKvs] = await Promise.all([
+      CSProxy.getAccount(baseUrl, account.address),
+      CSProxy.getAccountKvs(baseUrl, account.address),
+    ]);
+    const newAccount = accountToRawAccount(account, previousAccount as any, previousKvs as Kvs);
+
+    if (verbose) {
+      console.log('Setting account', newAccount);
+    }
+
+    const result = Proxy.fetch(
+      `${baseUrl}/simulator/set-state`,
+      [newAccount],
+    );
+
+    if (autoGenerateBlocks) {
+      await result;
+
+      await CSProxy.generateBlocks(baseUrl);
+    }
+
+    return result;
+  }
+
+  setAccount(account: EncodableAccount) {
+    return CSProxy.setAccount(this.baseUrl, account, this.autoGenerateBlocks, this.verbose);
+  }
+
+  async sendTx(tx: BroadTx) {
+    if (this.verbose) {
+      console.log('Sending transaction', tx);
+    }
+
+    const result = super.sendTx(tx);
+
+    if (this.autoGenerateBlocks) {
+      await result;
+
+      await this.generateBlocks();
+    }
+
+    return result;
+  }
+
+  static async getCompletedTxRaw(baseUrl: string, txHash: string) {
+    let res = await Proxy.getTxProcessStatusRaw(baseUrl, txHash);
+
+    let retries = 0;
+
+    while (!res || res.code !== 'successful' || res.data.status === 'pending') {
+      // We need delay since cross shard changes might not have been processed immediately
+      await new Promise((r) => setTimeout(r, 250));
+
+      if (res && res.data && res.data.status === 'pending') {
+        await CSProxy.generateBlocks(baseUrl);

```

Comment:
The Chain Simulator is not that reliable and sometimes multiple blocks need to be generated before the transaction is actually changed from `pending` state, hence this logic is needed.

If a developer wants greater control, they will set `autoGenerateBlocks` to fals and use the `sendTx` function directly instead, since that does not call `getCompleteTxRaw`. Also check a previous comment related to this.

Reply:
This logic should be moved CSWorld